patches/tModLoader/Terraria/ModLoader/PlayerDrawLayerHooks.cs

```diff
@@ -0,0 +1,99 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Terraria.DataStructures;
+using static Terraria.ModLoader.PlayerDrawLayer;
+
+namespace Terraria.ModLoader
+{
+	public static class PlayerDrawLayerHooks
+	{
+		private static List<PlayerDrawLayer> _layers = new List<PlayerDrawLayer>(PlayerDrawLayers.VanillaLayers);
+		public static IReadOnlyList<PlayerDrawLayer> Layers => _layers;
+
+		private static PlayerDrawLayer[] _drawOrder;
+		public static IReadOnlyList<PlayerDrawLayer> DrawOrder => _drawOrder;
+
+		internal static void Add(PlayerDrawLayer layer) => _layers.Add(layer);
+
+		internal static void Unload() {
+			_layers = new List<PlayerDrawLayer>(PlayerDrawLayers.VanillaLayers);
+			foreach (var layer in _layers) {
+				layer.ClearChildren();
+			}
+		}
+
+		internal static void ResizeArrays() {
+			var positions = Layers.ToDictionary(l => l, l => l.GetDefaultPosition());
+
+			PlayerHooks.ModifyDrawLayerOrdering(positions);
+
+			// net core please!!
+			//foreach (var (layer, pos) in positions.ToArray()) {
+			foreach (var kv in positions.ToArray()) {
+				var layer = kv.Key;
+				switch (kv.Value) {
+					case Between _:
+						continue;
+					case Before b:
+						b.Layer.AddChildBefore(layer);
+						break;
+					case After a:
+						a.Layer.AddChildAfter(layer);
+						break;
+					case Multiple m:
+						int slot = 0;
+						foreach (var (pos, cond) in m.Positions)
+							positions.Add(new MobilePlayerDrawLayerSlot(layer, cond, slot++), pos);
+						break;
+					default:
+						throw new ArgumentException($"PlayerDrawLayer {layer} has unknown Position type {kv.Value}");
```

Comment:
Can't we use expressive switch cases yet?

Reply:
Maybe? Link?

Reply:
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/switch-expression
See first example

Reply:
We can, but these are statements, not expressions.